lib/temporal/client.rb

```diff
@@ -0,0 +1,382 @@
+require 'json'
+require 'securerandom'
+require 'socket'
+require 'temporal/errors'
+require 'temporal/converter'
+require 'temporal/interceptor/client'
+require 'temporal/interceptor/chain'
+require 'temporal/workflow/handle'
+require 'temporal/workflow/execution_info'
+require 'temporal/workflow/id_reuse_policy'
+require 'temporal/workflow/query_reject_condition'
+
+module Temporal
+  class Client
+    attr_reader :namespace, :identity
+
+    # TODO: More argument to follow for converters, codecs, etc
+    def initialize(connection, namespace, interceptors: [])
+      @connection = connection
+      @namespace = namespace
+      @identity = "#{Process.pid}@#{Socket.gethostname}"
+      @interceptor_chain = Interceptor::Chain.new(interceptors)
```

Comment:
I did consider that as a solution, however I believe that this one is much better because it doesn't force the shape of the constructor method

Reply:
You don't have to enforce constructor shape, you can just have an init function. We have constructors that accept "next" in some other SDKs because their default implementations then do the right thing. It is important IMO for an interceptor to have its next interceptor and choose whether to invoke it.

Reply:
I'm not sure how is it different from this implementation. `super` or `yield` gives you the control over calling the next interceptor or not. You just don't need to know what that interceptor is

Reply:
If the result is the same, that's fine. All other SDKs accept the downstream interceptors, but I guess if it makes sense for Ruby interceptors to be unique here, ok. Doesn't make sense in my head to use a loop instead of actual method interception, but if API is the same, meh.

Question to confirm - can my interceptor report the time it takes for downstream interceptors to run? This is crucial to our tracing interceptors.

Reply:
Of course, yes:

```ruby
class LatencyInterceptor < Temporal::Interceptor::Client
  def start_workflow(input)
    start_time = Time.now
    result = super
    logger.log("Time elapsed: #{Time.now - start_time}")    
    result
  end
end
```